content/blog/2022-05-22-object/index.md

```diff
@@ -0,0 +1,162 @@
+
+## What I've actually done?
+In this project, I have designed a couple of new primitives, which provide higher-level abstractions. The whole implementation, is basically just a compiler pass which translates a Bril program with object-related syntax to a Bril program with only minimal/core syntax, so that it's completely modular and does not change the behavior of current Bril programs.
+
+## A class-based implementation
+A key question in the design is that whether to use a class-based object model or a prototype-based object model. In a class-based model, objects are pre-defined for their behaviors and types, meanwhile in a prototype-based model, objects are the primary entities, and objects pass all methods and values with each other through delegations. While a prototyped-based model is considered much simpler, I picked the class-based model since Bril is statically-typed and I don't want to add something that break the well-formedness of Bril. It also suits the current Bril toolings more and would get us rid of hectic dynamic typecheckings.
+
+## An example
+Let's start from an extended Bril program, which I would use to explain any new syntax and translation.
+
+``` 
+%rectangle {
+  length: int;
+  width: int:
+
+  @init(input_l: int, input_w: int){
+    lengthref: ref = dot this length;
+    widthref: ref = dot this width;
+
+    set lengthref input_l;
+    set widthref input_w;
+  }
+
+  @area: int{
+    lengthref: ref = dot this length;
+    widthref: ref = dot this width;
+
+    length: int = get lengthref;
+    width: int = get widthref;
+
+    prod: int = mul length width; 
+    ret prod;
+  }
+}
+
+@main {
+  a: int = const 4;
+  b: int = const 2;
+
+  rect1: rectangle = new rectangle;
+  initref: ref = dot rect1 init;
+  call @initref a b;
+  
+  arearef: ref = dot rect1 area;
+  area: int = call @arearef;
+  
+  print area;
+
+  delete rect1;
+}
+```
+
+This Bril program takes in two integers `a`, and `b`, uses them as `length` and `width` of a rectangle, and then outputs the `area` which is the `product` of the two variables. Compare to the world where we do not have classes and objects, this clearly has a lot of overhead in terms of extra instructions. However, we now clearly see that we are forming a rectangle with the given inputs, and we know we are computing the area of the rectagle we just formed.
+
+So what the translated Bril program with only core syntax would look like?
+
+```
+@rectangle.init(this.length.ptr: ptr<int>, this.width.ptr: ptr<int>, input_l: int, input_w: int) {
+  store this.length.ptr input_l;
+  store this.length.ptr input_w;
+}
+
+@rectangle.area(this.length.ptr: ptr<int>, this.width.ptr: ptr<int>) {
+  length: int = load this.length.ptr;
+  width: int = load this.width.ptr;
+
+  prod: int = mul length width; 
+  ret prod;
+}
+
+@main {
+  a: int = const 4;
+  b: int = const 2;
+  
+  one: int = const 1;
+  rect1.length.ptr: ptr<int> = alloc one;
+  rect1.width.ptr: ptr<int> = alloc one;
```

Comment:
This approach, where every object variable gets translated into a _series_ of variables corresponding to its fields, seems like a necessary consequence of the current memory extension. However, it seems to make a few things sort of complicated: like, for example, passing an object into a function. That would require changing the function signature to take multiple arguments in place of the original (class-typed) argument. I wonder if it wouldn't make sense to build a lower-level abstraction with heterogeneously-typed structs as an intermediate point.